### PR TITLE
Support multiple versions of .net core

### LINKED ISSUE #465: Support Dockerfile for .net core 2.2 apps

### PR DESCRIPTION
**Pull request by StephenWeatherford**
Fixes #465 

Sync up with what Visual Studio does for .NET core apps (console and asp.net).  I've taken their templates verbatim to make it easier to sync up with changes.

### CODE REVIEW

File: configureWorkspace/configure.ts [hunk 1]
```diff
@@ -415,8 +412,11 @@ async function configureCore(actionContext: IActionContext, options: ConfigureAp
 
         if (writeFile) {
             // Paths in the docker files should be relative to the Dockerfile (which is in the output folder)
-            fs.writeFileSync(filePath, generatorFunction(serviceNameAndPathRelativeToOutput, platformType, os, port, packageInfo), { encoding: 'utf8' });
-            filesWritten.push(fileName);
+            let fileContents = generatorFunction(serviceNameAndPathRelativeToOutput, platformType, os, port, packageInfo);
+            if (fileContents) {
+                fs.writeFileSync(filePath, fileContents, { encoding: 'utf8' });
+                filesWritten.push(fileName);
+            }
```
Comment (PrashanthCorp) [inline configureWorkspace/configure.ts]: This looks like a bug that would have affected other usecases too. Nice catch.

File: configureWorkspace/configure_dotnetcore.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    // example: COPY Core2.0ConsoleAppWindows/Core2.0ConsoleAppWindows.csproj Core2.0ConsoleAppWindows/
+    let copyProjectCommands = `COPY ["${serviceNameAndRelativePath}.csproj", "${projectDirectory}/"]`
+    let exposeStatements = `EXPOSE ${port}`;
+
+    // Parse version from TargetFramework
+    // Example: netcoreapp1.0
+    const defaultNetCoreVersion = '2.1';
+    let [netCoreAppVersion] = extractRegExGroups(version, /^netcoreapp([0-9.]+)$/, [defaultNetCoreVersion]);
+
+    // semver requires a patch in the version, so add it if only major.minor
```
Comment (PrashanthCorp) [inline configureWorkspace/configure_dotnetcore.ts]: Comment looks incomplete? did you mean `if version contains only major.minor`?

File: configureWorkspace/configure_dotnetcore.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            break;
+        case "ASP.NET Core":
+            template = os === "Linux" ? aspNetCoreLinuxTemplate : aspNetCoreWindowsTemplate;
+            break;
+        default:
+            throw new Error(`Unexpected platform "${platform}"`);
+    }
+
+    let contents = template.replace('$base_image_name$', baseImageName)
+        .replace(/\$expose_statements\$/g, exposeStatements)
```
Comment (PrashanthCorp) [inline configureWorkspace/configure_dotnetcore.ts]: It it not clear to me whether #391 would be handled. There is no conditional logic in inserting the `$expose_statements$`. It is present b default on both aspnetCore templates,  and absent by default on both dotnetCoreTemplates

File: commands/utils/docker-endpoint.ts [hunk 4]
```diff
@@ -117,12 +118,20 @@ class DockerClient {
...
-                const ports = Object.keys(exposedPorts);
-                resolve(ports);
+    public async getExposedPorts(imageId: string): Promise<string[]> {
+        return await new Promise<string[]>((resolve, reject) => {
+            this.getImage(imageId).inspect((error: {}, data: { Config: { ExposedPorts: {} } }) => {
+                if (error) {
+                    reject(error);
+                } else {
+                    let exposedPorts = data.Config.ExposedPorts;
+                    if (!exposedPorts) {
```
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply commands/utils/docker-endpoint.ts]: True.  Not worth changing now that it's been tested, though.

File: configureWorkspace/configure.ts (hunk 1 above)
Comment (StephenWeatherford) [reply configureWorkspace/configure.ts]: You mean if (fileContents)?
This is actually just a change in behavior - moving special-case code out of the common code.

File: configureWorkspace/configure_dotnetcore.ts (hunk 3 above)
Comment (StephenWeatherford) [reply configureWorkspace/configure_dotnetcore.ts]: It's always present in the template now, but exposeStatements will be empty string if no port.  As you see, I handled this more thoroughly for other platforms in https://github.com/Microsoft/vscode-docker/pull/490 but don't plan to merge until after shipping.

Comment (StephenWeatherford) [reply configureWorkspace/configure_dotnetcore.ts]: (by now, I mean with the PR that came after this)

Comment (PrashanthCorp) [reply configureWorkspace/configure_dotnetcore.ts]: Agreed. Thanks for the fix there! I was just pointing out a bug we triaged.

**Pull request merged**